### PR TITLE
Best effort granularity comparing

### PR DESCRIPTION
**Pull request by navis**
In some cases, if granularity can be compared with each other, some works can be avoided. For example, if group-by query is done by 1H, we need not to recalculate ordering granularity of 10M, 30M, etc. (we should calculate if it's 45M or 1D, of course)

### CODE REVIEW

**Comment by himanshug** [comment]
couldn't understand what additional code change will use this PR, can you make the other changes as well which need this?

**Comment by navis** [comment]
I'm doing it on #2032. I'll merge that if it's better.

**Comment by himanshug** [comment]
ok just looked at the other PR, I think merging is good idea so that you can use the method from here for real and it will be clearer.

File: processing/src/main/java/io/druid/granularity/QueryGranularity.java [hunk 1]
```diff
@@ -86,4 +89,74 @@ else if(o instanceof Number)
     }
     throw new IAE("Cannot convert [%s] to QueryGranularity", o.getClass());
   }
+
+  // best effort comparing
+  public static Integer compare(QueryGranularity g1, QueryGranularity g2) {
```
Comment (himanshug) [inline processing/src/main/java/io/druid/granularity/QueryGranularity.java]: it would be better to have 
```
public abstract Integer compare(QueryGranularity granularity);
```
and each implementation implementing it, rather than having it all in one place for all concrete types.

Comment (navis) [inline processing/src/main/java/io/druid/granularity/QueryGranularity.java]: @himanshug Thought on that once. But I wanted make it clear that it's just best effort try. Anyway, I'll do that.

**Comment by navis** [comment]
@himanshug addressed comment

**Comment by fjy** [comment]
I read this over and I think generally it is fine. There's some minor style problems and probably an issue with your auto-formatting. Or a bug with our style guides.

**Comment by fjy** [comment]
:+1:

**Comment by navis** [comment]
@fjy Reformatted. Sorry.

File: processing/src/main/java/io/druid/granularity/QueryGranularity.java [hunk 2]
```diff
@@ -88,4 +88,23 @@ else if(o instanceof Number)
...
+      return 0;
+    }
+    if (d1 > d2 && (d1 % d2 == 0)) {
+      return 1;
+    }
+    if (d2 > d1 && (d2 % d1 == 0)) {
+      return -1;
+    }
+    return null;
+  }
```
Comment (himanshug) [inline processing/src/main/java/io/druid/granularity/QueryGranularity.java]: can you move this to DurationGraunlarity?

Comment (himanshug) [inline processing/src/main/java/io/druid/granularity/QueryGranularity.java]: nvmd, i see it is used in multiple places

File: processing/src/main/java/io/druid/granularity/PeriodGranularity.java [hunk 3]
```diff
@@ -330,6 +321,43 @@ private long truncateMillisPeriod(final long t)
...
+      Period p2 = pg2.getPeriod();
+      if (!isPrecise(p1, c1) || !isPrecise(p2, c2)) {
+        return null;
+      }
+      return compareMultiplicity(p1.toStandardDuration().getMillis(), p2.toStandardDuration().getMillis());
+    }
+    return super.compare(other);
+  }
+
+  private boolean isPrecise(Period period, Chronology chronology)
```
Comment (himanshug) [inline processing/src/main/java/io/druid/granularity/PeriodGranularity.java]: can you please explain or add a comment regarding what this ensures?

Comment (navis) [inline processing/src/main/java/io/druid/granularity/PeriodGranularity.java]: I've copied javadoc of `DruationField#isPrecise()`. Time related things are really confusing.

File: processing/src/main/java/io/druid/granularity/PeriodGranularity.java [hunk 4]
```diff
@@ -306,7 +314,7 @@ private long truncateMillisPeriod(final long t)
   {
     // toStandardDuration assumes days are always 24h, and hours are always 60 minutes,
     // which may not always be the case, e.g if there are daylight saving changes.
-    if (chronology.days().isPrecise() && chronology.hours().isPrecise()) {
+    if (isPrecise()) {
```
Comment (xvrl) [inline processing/src/main/java/io/druid/granularity/PeriodGranularity.java]: can you explain why this change is needed?

File: processing/src/main/java/io/druid/granularity/QueryGranularity.java [hunk 5]
```diff
@@ -88,4 +88,23 @@ else if(o instanceof Number)
     }
     throw new IAE("Cannot convert [%s] to QueryGranularity", o.getClass());
   }
+
+  // best effort comparing
+  public Integer compare(QueryGranularity other) {
```
Comment (xvrl) [inline processing/src/main/java/io/druid/granularity/QueryGranularity.java]: can we give a clear definition of what "comparing" means?

File: processing/src/main/java/io/druid/granularity/PeriodGranularity.java (hunk 4 above)
Comment (navis) [inline processing/src/main/java/io/druid/granularity/PeriodGranularity.java]: If all period used in PeriodGranualrity are composed with days and hours, it's ok. But if it's not (week, month, year), we should make it sure that all those fields are precise, too. And if it's just composed with seconds or minutes, we don't need to bother whatever days and hours are.

File: processing/src/main/java/io/druid/granularity/QueryGranularity.java (hunk 5 above)
Comment (navis) [inline processing/src/main/java/io/druid/granularity/QueryGranularity.java]: With a given interval and granularity, we can make a set of truncated timestamps. If this time-set is a subset of other time-set made from some granularity with the same interval, we might say that the former granularity is bigger than the later and vise-versa. 

I've thought at first that this could be used for some optimizations but it seemed hard to find meaningful situation for using it. Any thoughts?

**Comment by fjy** [comment]
@navis can you provide some benchmarks around what kind of performance improvement this is giving?
